### PR TITLE
Removed duplicate code in python driver

### PR DESCRIPTION
**Pull request by timmartin19**
The python driver contained an absurd amount of duplicate code involving find_element_\* and find_elements_*. By putting it in an Abstract Base Class we can ensure that any class that claims to use them fully implements it. Furthermore, this makes editing how they are performed very easy. For example, if a new type is added, you only have to put it in one place. Using an abstract base class means that if the find_element and find_elements methods are not implemented, an exception will be raised on import further ensuring safe use. Finally, it makes extension significantly easier. Although this could be further improved using injection rather than inheritance.

tl:dr; dried out some WET code.

### CODE REVIEW

**Comment by lukeis** [comment]
why were you removing the intellij files with this pull request? We do actually like to have those in the repo.

This is a lot to chew in one request, it would be nice to separate concerns in different pull requests, also could you please squash the commits for each of those concerns.

Also it appears the changes are a bit stale and have conflicts.

**Comment by andreastt** [comment]
What @lukeis said.  Please resubmit in smaller PRs.
